components/scream/src/physics/p3/p3_functions_f90.cpp

```diff
@@ -1728,6 +1748,24 @@ Real cxx_pow(Real base, Real exp)
 #endif
 }
 
+Real cxx_square(Real input)
+{
+#ifdef KOKKOS_ENABLE_CUDA
+  return CudaWrap<Real, DefaultDevice>::cxx_square(input);
+#else
+  return input*input;
+#endif
+}
+
+Real cxx_cube(Real input)
+{
+#ifdef KOKKOS_ENABLE_CUDA
+  return CudaWrap<Real, DefaultDevice>::cxx_cube(input);
```

Comment:
You don't need this wrapping call. A wrapper to a kernel dispatch is used when a function like pow can diff in n ULP on GPU vs CPU. In this case, the implementation is x*x*x, which is exact in a debug build across architectures.

Reply:
You could also just impl this in F90, since you're just making clear the impl is `x*x*x`. Just make an F90 function called `cube` and use it instead of `x**3`. That will match the C++ `cube`. Again, this is a different situation than that of, e.g., `pow`. Here we're pinning down the underlying meaning exactly. With `pow` we're working around library diffs.

Reply:
Thanks for the feedback, @ambrad . I think implementing these in F90 is a good idea. What's the best place to put these functions? `components/scream/src/physics/p3/micro_p3_iso_f.f90`? So far it looks like that file just contains an interface for the Fortran/C++ bridge.

Reply:
Alternatively, what about just `#define`ing `bfb_cube(x)` as `(x)*(x)*(x)` and `bfb_square` as `(x)*(x)`, pulling those macros out of the `#else` portion of `#ifdef SCREAM_CONFIG_IS_CMAKE`? That defines the expressions in all cases.

Reply:
Either is fine. You can put these defs or functions in micro_p3.F90 itself.